Create design-update.md (#40)
sensein/etelemetry-client: docs/design-update.md

# Design doc to facilitate feature specification

## Etelemetry service objectives:

> - Return useful information to the user about the state of the software (bugs, new versions, etc.,.)
> - Track software usage

Reviewer: ```suggestion
- Track software usage. Annotations:
  - Process starts
  - Process ends + exit status
  - (possibly) Elapsed time
  - (possibly) Memory utilization
  - (possibly) CPU utilization
- Feedback to users:
  - Notice of new versions available
  - "Retired" versions (i.e., bad versions with disqualifying bugs)
```

# Design doc to facilitate feature specification

## Etelemetry service objectives:
- Track software usage. Annotations:
  - Process starts
  - Process ends + exit status
  - Heartbeats for long runtime tasks
  - Software version
  - Platform (bare-metal, container, etc.)
  - Relevant options
  - "Session" information
  - "User" information

>   - (possibly) Elapsed time

Reviewer: ```suggestion
  - (possibly) Input dataset hash - so that we could identify open data executions (if the dataset is run unaltered).
  - (possibly) Elapsed time
```

  - (possibly) Memory utilization
  - (possibly) CPU utilization
- Feedback to users:
  - Notice of new versions available
  - "Retired" versions (i.e., bad versions with disqualifying bugs)
  - ETA (progress bar or such, using previously acquired stats from other users)

## Wishlist

> - use a more generic logging tool as the backend (like prometheus)

Reviewer: ```suggestion
- identification logic to establish user/session constructs that allow: 1) defining a "unique users" concept; 2) defining "exit points" where sessions are aggregated (i.e., 90% of the processes ended on this particular error).
- use a more generic logging tool as the backend (like prometheus)
```

- discard IP addresses after geo conversion
- define the message packet, the transformed message packet
- thus far we have been able to do things without caring who used the service 
  (even though not advertised). if messages become arbitrary, we may need 
  registration/auth tokens. brings up a whole different level of complexity that 
  i would like to avoid.

> - make querying, accessing, and doing analytic plots

Reviewer 2: +1 to making the data acquired more usable, just need to be mindful of large server load increase

- spin up your own backend on heroku/some such

> 

Reviewer 2: ```suggestion
- revisit backend DB choice (mongo/sqlite/redis)
```

## Message packets

> 

Reviewer: ```suggestion
- Try TCP/Restful first, resort to UDP if communication fails
```
